tcms/utils/permissions.py

```diff
@@ -3,30 +3,39 @@
 from django.contrib.auth.models import Group, Permission
 
 
-def assign_default_group_permissions():
+def assign_default_group_permissions(output=None, refresh_all=False):
     """
         Adds the default permissions for Administrator and Tester
         groups!
     """
     admin = Group.objects.get(name='Administrator')
-    if admin.permissions.count() == 0:
-        all_perms = Permission.objects.all()
-        admin.permissions.add(*all_perms)
+    if admin.permissions.count() == 0 or refresh_all:
+        perms_to_add = Permission.objects.all().exclude(
+            pk__in=admin.permissions.all())
+        admin.permissions.add(*perms_to_add)
+        if output:
+            for perm in perms_to_add:
+                output.write('%s.%s added to %s group' %
+                             (perm.content_type.app_label, perm.codename,
+                              admin.name))
 
     tester = Group.objects.get(name='Tester')
-    if tester.permissions.count() == 0:
+    tester_perms = tester.permissions.all()
+    if tester_perms.count() == 0 or refresh_all:
         # apply all permissions for test case & product management
         for app_name in ['bugs', 'django_comments', 'linkreference', 'management',
-                         'testcases', 'testplans', 'testruns']:
-            app_perms = Permission.objects.filter(content_type__app_label__contains=app_name)
+                         'testcases', 'testplans', 'testruns', 'attachments']:
+            app_perms = Permission.objects.filter(
+                content_type__app_label__contains=app_name)
+            app_perms = app_perms.exclude(pk__in=tester_perms).exclude(
+                content_type__app_label='attachments',
+                codename='delete_foreign_attachments')
             tester.permissions.add(*app_perms)
-
-    # this app was introduced later and we don't want all of its permissions
-    if tester.permissions.filter(content_type__app_label='attachments').count() == 0:
-        attachment_perms = Permission.objects.filter(
-            content_type__app_label='attachments'
-        ).exclude(codename='delete_foreign_attachments')
-        tester.permissions.add(*attachment_perms)
+            if output:
+                for perm in app_perms:
+                    output.write('%s.%s added to %s group' %
+                                 (perm.content_type.app_label, perm.codename,
+                                  tester.name))
```

Comment:
This block is duplicated here and above. Move it into a small function. Can stay in the same module.